### DISCUSSION TITLE
An idea to optimize the full refresh of epaper useing GxEPD2

### DISCUSSION THREAD

**Discussion by pangqianjin** [ZinggJM/GxEPD2#70] [Ideas]
Hello! I noticed that there is a private variable named "_buffer" at the source code, which was used to send the prepared  contents to the screen. And I just had a thought that I want to access this "_buffer" 's value, when I need to fully updated my epaper after 5~6 partial updates(which will clear the shadow). I' d like to copy this _buffer firstly before the full update, and after full updating I'd like to use display.epd2.writeImage(_copied_buffer, 0, 0, GxEPD2_DRIVER_CLASS::WIDTH, GxEPD2_DRIVER_CLASS::HEIGHT) to re-draw the screen using the data before full updating.

However the "_buffer" is a private variable and I have not found some ways to do this.

**Comment by G6EJD** [comment]
Take private out of the definition for your implementation.

**Reply by pangqianjin** [reply]
Thanks for your apply！

**Comment by ZinggJM** [comment]
Accessing member variables directly is a bad design (except for const attributes).
Private members are usually declared _private by design, to avoid side effects by bad use.

GxEPD2 transposes partial windows to (0,0) in the buffer, so your use would not work.
The buffer may be smaller than the buffer in the controller, when using paged display.

**Reply by pangqianjin** [reply]
Thanks to your reply, I am trying another way to do this.

**Reply by ZinggJM** [reply]
Adafruit_GFX has canvas for this.
```
/// A GFX 1-bit canvas context for graphics
class GFXcanvas1 : public Adafruit_GFX {
```
